interoptest/src/goservice/testservice/grpc_receiver.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2018, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package testservice
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/census-ecosystem/opencensus-experiments/interoptest/src/goservice/genproto"
+	"google.golang.org/grpc"
+)
+
+// GrpcReceiver is the type that handles test requests over GRPC.
+type GrpcReceiver struct {
+	mu     sync.Mutex
+	ln     net.Listener
+	server *grpc.Server
+
+	receiver *GrpcTestReceiver
+
+	stopOnce              sync.Once
+	startServerOnce       sync.Once
+	startRegistrationOnce sync.Once
+}
+
+var (
+	errAlreadyStarted = errors.New("already started")
+	errAlreadyStopped = errors.New("already stopped")
+)
+
+// New just creates the test services for request over GRPC.
+func New(addr string) (*GrpcReceiver, error) {
+	// TODO: consider using options.
+	ln, err := net.Listen("tcp", addr)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to bind to address %q: error: %v", addr, err)
+	}
+	h := &GrpcReceiver{ln: ln}
+
+	return h, nil
+}
+
+func (gr *GrpcReceiver) registerTestReceiver() error {
+	var err = errAlreadyStarted
+
+	gr.startRegistrationOnce.Do(func() {
+		gr.receiver = &GrpcTestReceiver{registeredServices: make(map[string][]*interop.TestRequest)}
+		srv := gr.grpcServer()
+		interop.RegisterTestExecutionServiceServer(srv, gr.receiver)
+		err = nil
+	})
+
+	return err
+}
+
+func (gr *GrpcReceiver) grpcServer() *grpc.Server {
+	gr.mu.Lock()
+	defer gr.mu.Unlock()
+
+	if gr.server == nil {
+		gr.server = grpc.NewServer()
+	}
+
+	return gr.server
+}
+
+// Start runs the test service.
+func (gr *GrpcReceiver) Start(ctx context.Context) error {
+	if err := gr.registerTestReceiver(); err != nil && err != errAlreadyStarted {
+		return err
+	}
+
+	if err := gr.startGRPCServer(); err != nil && err != errAlreadyStarted {
+		return nil
```

Comment:
This should be `return err`. (see also https://github.com/census-instrumentation/opencensus-service/pull/264)